src/object/cli_obj.c

```diff
@@ -4910,6 +5193,91 @@ obj_retry_next_shard(struct dc_object *obj, struct obj_auxi_args *obj_auxi,
 	return 0;
 }
 
+/* some pre-process for cond_fetch -
+ * 1. create local TX if it is no in a user transaction - to ensure all the fetches on same epoch
+ *    chosen by first operation (reuse dtx_epoch_chosen related logic).
+ * 2. for multiple-akeys case, split obj task to multiple sub-tasks each for one akey. For this
+ *    case return 1 to indicate wait sub-tasks' completion.
+ */
+static int
+obj_cond_fetch_prep(tse_task_t *task, struct obj_auxi_args *obj_auxi)
+{
+	daos_obj_fetch_t	*args = dc_task_get_args(task);
+	d_list_t		*task_list = &obj_auxi->shard_task_head;
+	tse_task_t		*sub_task;
+	d_sg_list_t		*sgl;
+	bool			 per_akey = args->flags & DAOS_COND_PER_AKEY;
+	uint64_t		 fetch_flags;
+	uint32_t		 i;
+	int			 rc = 0;
+
+	if (!daos_handle_is_valid(obj_auxi->th)) {
+		D_ASSERT(!obj_auxi->tx_local);
+		D_ASSERT(!daos_handle_is_valid(args->th));
+		rc = dc_tx_local_open(obj_auxi->obj->cob_coh, 0, 0, &obj_auxi->th);
+		if (rc) {
+			D_ERROR("task %p "DF_OID" dc_tx_local_open failed "
+				DF_RC"\n", task, DP_OID(obj_auxi->obj->cob_md.omd_id),
+				DP_RC(rc));
+			return rc;
+		}
+		args->th = obj_auxi->th;
+		obj_auxi->tx_local = 1;
+	}
+
+	if (args->nr <= 1 || (args->flags & (DAOS_COND_AKEY_FETCH | DAOS_COND_PER_AKEY)) == 0)
+		return rc;
+
+	/* If cond_fetch include multiple akeys, splits the obj task to multiple sub-tasks, one for
+	 * each akey. Because -
+	 * 1. for each akey's cond_fetch if any shard returns 0 (exist) then the akey is exist.
+	 * 2. for multi-akeys' cond_fetch, should return non-exist if any akey non-exist.
+	 * Now one fetch request only with one return code. So creates one sub-task for each akey.
+	 */
+	D_ASSERT(d_list_empty(task_list));
+	D_ASSERT(obj_auxi->cond_fetch_split == 0);
+	for (i = 0; i < args->nr; i++) {
+		fetch_flags = per_akey ? args->iods[i].iod_flags : args->flags;
+		sgl = args->sgls != NULL ? &args->sgls[i] : NULL;
+		rc = dc_obj_fetch_task_create(args->oh, obj_auxi->th, fetch_flags, args->dkey, 1,
+					      0, &args->iods[i], sgl, NULL, NULL, NULL,
+					      NULL, tse_task2sched(task), &sub_task);
+		if (rc) {
+			D_ERROR("task %p "DF_OID" dc_obj_fetch_task_create failed, "DF_RC"\n",
+				task, DP_OID(obj_auxi->obj->cob_md.omd_id), DP_RC(rc));
+			goto out;
+		}
+
+		tse_task_addref(sub_task);
+		tse_task_list_add(sub_task, task_list);
+
+		rc = dc_task_depend(task, 1, &sub_task);
+		if (rc) {
+			D_ERROR("task %p "DF_OID" dc_task_depend failed "DF_RC"\n",
+				task, DP_OID(obj_auxi->obj->cob_md.omd_id), DP_RC(rc));
+			goto out;
+		}
+
+		D_DEBUG(DB_IO, DF_OID" created sub_task %p for obj task %p\n",
+			DP_OID(obj_auxi->obj->cob_md.omd_id), sub_task, task);
+	}
+
+out:
+	if (rc == 0) {
+		D_DEBUG(DB_IO, "scheduling %d sub-tasks for cond_fetch IO task %p.\n",
+			args->nr, task);
+		obj_auxi->no_retry = 1;
```

Comment:
it just because it need not retry, because it is parent task just wait all its sub-tasks' completion. sub-task will be retried when needed.

Reply:
ah, these are per akey subtask. I see. thanks.  Though I really think these sub-tasks cause unnecessary complexity.

Reply:
I just though changing RPC probably is more complex, let's keep that possibility later, maybe in future we have some other stronger reason to change fetch RPC's reply, we can refine that together.

Reply:
hmm, I think we can use iod_size to indicate if the akey exist?  or even add -1 to iod_size for each akey?

Reply:
hmm, using orw_iod_sizes is considerable. But i am not very sure if that also is a kind of interoperability issue, because it requires the same version of client/server for example new client cannot work with old server as old server don't handle that. can think more later.